### PR TITLE
SONARUCFG-7 add newObject method

### PR DESCRIPTION
**Pull request by benzonico**

### CODE REVIEW

File: src/main/java/org/sonar/ucfg/Instruction.java [hunk 1]
```diff
@@ -54,6 +55,51 @@ private Terminator(InstructionType type, @Nullable LocationInFile locationInFile
...
+  public static class NewObject extends Instruction {
+    private final Expression.Variable lhs;
+    private final String instanceType;
+
+    private final int hash;
+    public NewObject(LocationInFile locationInFile, Expression.Variable lhs, String instanceType) {
+      super(InstructionType.NEW, locationInFile);
+      this.lhs = lhs;
+      this.instanceType = instanceType;
+      this.hash = Objects.hash(lhs, instanceType, locationInFile);
```
Comment (ghost) [inline src/main/java/org/sonar/ucfg/Instruction.java:67]: Just wondering now, but why not include the `InstructionType` into the hash as well? (same comment for other instructions.)

File: src/main/java/org/sonar/ucfg/Instruction.java [hunk 2]
```diff
@@ -54,6 +55,51 @@ private Terminator(InstructionType type, @Nullable LocationInFile locationInFile
     }
   }
 
+  public static class NewObject extends Instruction {
+    private final Expression.Variable lhs;
+    private final String instanceType;
+
+    private final int hash;
```
Comment (ghost) [inline src/main/java/org/sonar/ucfg/Instruction.java]: separate this line from the constructor

File: src/main/java/org/sonar/ucfg/BasicBlock.java [hunk 3]
```diff
@@ -56,16 +56,16 @@ public Label label() {
     return label;
   }
 
-  public List<Instruction.AssignCall> calls() {
-    return calls;
+  public List<Instruction> calls() {
```
Comment (ghost) [inline src/main/java/org/sonar/ucfg/BasicBlock.java]: should we rename this method as well?

File: src/main/java/org/sonar/ucfg/BasicBlock.java [hunk 4]
```diff
@@ -45,9 +45,9 @@ public void updateSuccs(Set<BasicBlock> successors) {
     }
   };
 
-  BasicBlock(Label label, List<Instruction.AssignCall> calls, Instruction.Terminator terminator, @Nullable LocationInFile loc) {
+  BasicBlock(Label label, List<Instruction> instructions, Instruction.Terminator terminator, @Nullable LocationInFile loc) {
     this.label = label;
-    this.calls = calls;
+    this.instructions = instructions;
```
Comment (ghost) [inline src/main/java/org/sonar/ucfg/BasicBlock.java:50]: I know that using the `BlockBuilder` we can only add NewObject or AssignCall. But if this constructor is called from elsewhere, I think we should validate the content of the `instructions` list here, to guarantee that there will be only `Instruction.AssignCall` or `Instruction.NewObject`.

File: src/main/java/org/sonar/ucfg/UCFGBuilder.java [hunk 5]
```diff
@@ -82,15 +82,24 @@ public BlockBuilder assignTo(Expression.Variable var, CallBuilder callBuilder) {
...
-      calls.add(new Instruction.AssignCall(loc, var, callBuilder.methodId, callBuilder.arguments));
+      instructions.add(new Instruction.AssignCall(loc, var, callBuilder.methodId, callBuilder.arguments));
+      return this;
+    }
+
+    public BlockBuilder newObject(Expression.Variable var,  String instanceType) {
+      return newObject(var, instanceType, LOC);
+    }
+
+    public BlockBuilder newObject(Expression.Variable var,  String instanceType, LocationInFile loc) {
```
Comment (ghost) [inline src/main/java/org/sonar/ucfg/UCFGBuilder.java]: double spacing between `var,  String`

File: src/main/java/org/sonar/ucfg/UCFGBuilder.java [hunk 6]
```diff
@@ -82,15 +82,24 @@ public BlockBuilder assignTo(Expression.Variable var, CallBuilder callBuilder) {
     }
 
     public BlockBuilder assignTo(Expression.Variable var, CallBuilder callBuilder, LocationInFile loc) {
-      calls.add(new Instruction.AssignCall(loc, var, callBuilder.methodId, callBuilder.arguments));
+      instructions.add(new Instruction.AssignCall(loc, var, callBuilder.methodId, callBuilder.arguments));
+      return this;
+    }
+
+    public BlockBuilder newObject(Expression.Variable var,  String instanceType) {
```
Comment (ghost) [inline src/main/java/org/sonar/ucfg/UCFGBuilder.java]: double spacing between `var,  String`

File: src/main/java/org/sonar/ucfg/UCFGtoProtobuf.java [hunk 7]
```diff
@@ -67,11 +72,17 @@ public static void toProtobufFile(UCFG ucfg, String filename) throws IOException
...
       .setVariable(assignCall.getLhs().id())
       .setMethodId(assignCall.getMethodId())
       .addAllArgs(assignCall.getArgExpressions().stream().map(UCFGtoProtobuf::toProtobuf).collect(Collectors.toList()))
-      .build();
+      .build()).build();
+  }
+
+  private static Ucfg.Instruction toProtobuf(Instruction.NewObject newObject) {
+    return Ucfg.Instruction.newBuilder().setNewObject(Ucfg.NewObject.newBuilder().setLocation(toProtobuf(newObject.location()))
+      .setType(newObject.instanceType())
```
Comment (ghost) [inline src/main/java/org/sonar/ucfg/UCFGtoProtobuf.java]: what about the `lhs` ? Aren't we supposed to also store the destination?

File: src/test/java/org/sonar/ucfg/UCFGtoProtobufTest.java [hunk 8]
```diff
@@ -43,6 +43,7 @@ void serialize_and_deserialize_ucfg() throws IOException {
     UCFGBuilder ucfgBuilder = UCFGBuilder.createUCFGForMethod("myMethod").addMethodParam(parameter1);
     Expression.Variable var1 = UCFGBuilder.variableWithId("var1");
     ucfgBuilder.addStartingBlock(newBasicBlock("startLabel", null)
+      .newObject(var1, "java.lang.Object", new LocationInFile("fileKey", 1,1, 1,12))
```
Comment (ghost) [inline src/test/java/org/sonar/ucfg/UCFGtoProtobufTest.java:46]: and no assertions?!

**Review by ghost — CHANGES_REQUESTED** [review]
Please have a look at my comments.

File: src/main/java/org/sonar/ucfg/BasicBlock.java (hunk 4 above)
Comment (benzonico) [reply src/main/java/org/sonar/ucfg/BasicBlock.java:50]: As discussed this should be part of a refactoring outside of this PR. see SONARUCFG-12

File: src/main/java/org/sonar/ucfg/Instruction.java (hunk 1 above)
Comment (benzonico) [reply src/main/java/org/sonar/ucfg/Instruction.java:67]: should be done in another ticket

**Review by ghost — APPROVED** [review]

**Pull request merged**